cloud/scope/powervs_cluster.go

```diff
@@ -882,37 +900,110 @@ func (s *PowerVSClusterScope) ReconcileNetwork() (bool, error) {
+		if err != nil {
+			return nil, err
+		}
+		if network == nil || network.NetworkID == nil {
+			s.V(3).Info("Unable to find PowerVS network in IBM Cloud", "network", s.IBMPowerVSCluster.Spec.Network)
+			return nil, nil
+		}
+		if err := s.verifyDHCPServerRespectToNetwork(*network.NetworkID); err != nil {
+			return nil, err
+		}
 		return network.NetworkID, nil
 	}
 
+	if s.IBMPowerVSCluster.Spec.Network.Name != nil {
+		networkID, err := checkNetworkByName(*s.IBMPowerVSCluster.Spec.Network.Name)
+		if err != nil {
+			return nil, err
+		}
+		return networkID, nil
+	}
+
+	if s.DHCPServer() != nil && s.DHCPServer().ID != nil {
+		dhcpServer, err := s.IBMPowerVSClient.GetDHCPServer(*s.DHCPServer().ID)
+		if err != nil {
+			return nil, err
+		}
+		networkID, err := checkNetworkByID(*dhcpServer.Network.ID)
+		if err != nil {
+			return nil, err
+		}
+		return networkID, nil
+	}
+
 	// if the user has provided the already existing dhcp server name then there might exist network name
 	// with format DHCPSERVER<DHCPServer.Name>_Private , try fetching that
 	var networkName string
 	if s.DHCPServer() != nil && s.DHCPServer().Name != nil {
-		networkName = fmt.Sprintf("DHCPSERVER%s_Private", *s.DHCPServer().Name)
+		networkName = dhcpNetworkName(*s.DHCPServer().Name)
 	} else {
 		networkName = *s.GetServiceName(infrav1beta2.ResourceTypeNetwork)
 	}
-
-	s.V(3).Info("Checking if PowerVS network exists in IBM Cloud with name", "name", networkName)
-	// fetch the network associated with name
-	network, err := s.IBMPowerVSClient.GetNetworkByName(networkName)
+	networkID, err := checkNetworkByName(networkName)
 	if err != nil {
 		return nil, err
 	}
-	if network == nil || network.NetworkID == nil {
-		s.V(3).Info("Unable to find PowerVS network in IBM Cloud", "network", s.IBMPowerVSCluster.Spec.Network)
-		return nil, nil
+
+	return networkID, nil
+}
+
+// verifyDHCPServerRespectToNetwork verifies DHCP server exists and it is active for the given network ID.
+func (s *PowerVSClusterScope) verifyDHCPServerRespectToNetwork(networkID string) error {
+	dhcpServers, err := s.IBMPowerVSClient.GetAllDHCPServers()
+	if err != nil {
+		return err
+	}
+	for _, dhcpServer := range dhcpServers {
+		if *dhcpServer.Network.ID == networkID {
+			if *dhcpServer.Status == string(infrav1beta2.DHCPServerStateActive) {
+				return nil
+			}
+			return fmt.Errorf("dhcp server with given network id found but it is not in active state, current status: %s", *dhcpServer.Status)
+		}
+	}
+	return fmt.Errorf("dhcp server with given network id '%s' not found", networkID)
+}
+
+// isNetworkAvailable verifies the user provided network ID is available and it is managed by DHCP.
+func (s *PowerVSClusterScope) isNetworkAvailable() (bool, error) {
+	nw, err := s.IBMPowerVSClient.GetNetworkByID(*s.GetNetworkID())
+	if err != nil {
+		return false, err
 	}
-	return network.NetworkID, nil
+
+	if nw != nil && !nw.DhcpManaged {
```

Comment:
Do you think nw.DhcpManaged is required? We are getting network id from status and we set network id to status only when we create DHCP server.

Reply:
Ok I can remove this check.